### PR TITLE
Warn on questionable (dis)use of `# Automatic Code`

### PR DESCRIPTION
**Pull request by madig**
1. Warn about mis-cased `# Automatic Code` comments (see #701)
2. Warn if there is no insertion marker but kerning.plist data and a manual kern/dist feature in the fea code (because kerning.plist will be dropped)

### CODE REVIEW

**Comment by anthrotype** [comment]
@madig are you going to finish this? i'd like to cut a release soon-ish

**Comment by madig** [comment]
Yes, one day. Currently looking at something else but hopefully this week.

File: Lib/ufo2ft/featureCompiler.py [hunk 1]
```diff
@@ -354,3 +369,32 @@ def buildTables(self):
...
+            if hasattr(statement, "statements"):
+                _inner(statement)
+            elif isinstance(statement, ast.Comment):
+                for pattern_case, pattern_ignore_case in patterns_compiled:
+                    match_case = re.match(pattern_case, str(statement))
+                    match_ignore_case = re.match(pattern_ignore_case, str(statement))
+                    if match_ignore_case and not match_case:
+                        logger.warning(
+                            "%s: The insertion comment '%s' in the feature file is "
+                            "miscased, ignoring it.",
```
[the remark this answers is not in the document]
Comment (madig) [reply Lib/ufo2ft/featureCompiler.py:395]: I wanted to do this at first, but each feature writer can dynamically supply their own marker patterns, which are regexes. I guess I could show the regex pattern to the user, but I'm not sure that's a good idea?

File: Lib/ufo2ft/util.py [hunk 2]
```diff
@@ -612,3 +612,15 @@ def unicodeScriptExtensions(
...
+def describe_ufo(ufo: Any) -> str:
+    """Returns a description of a UFO suitable for logging."""
+    if (
+        hasattr(ufo, "reader")
+        and hasattr(ufo.reader, "fs")
+        and hasattr(ufo.reader.fs, "root_path")
+    ):
+        return ufo.reader.fs.root_path
+    else:
+        return f"{ufo.info.familyName} {ufo.info.styleName}"
```
Comment (anthrotype) [inline Lib/ufo2ft/util.py:626]: these can both be unset or None, maybe if you end up with an empty string just use its repr(ufo)

File: Lib/ufo2ft/featureCompiler.py [hunk 3]
```diff
@@ -354,3 +369,32 @@ def buildTables(self):
...
+    patterns_compiled = tuple(
+        (re.compile(pattern), re.compile(pattern, re.IGNORECASE))
+        for pattern in patterns
+    )
+
+    def _inner(block: Any) -> None:
+        for statement in block.statements:
+            if hasattr(statement, "statements"):
+                _inner(statement)
+            elif isinstance(statement, ast.Comment):
```
Comment (anthrotype) [inline Lib/ufo2ft/featureCompiler.py:388]: I don't think it's a good idea to traverse the entire FEA tree for _any_ comments that match the pattern, we should only look within the statements that comprise a FeatureBlock, not anywhere else. I think that's where we look for these.

Comment (anthrotype) [reply Lib/ufo2ft/featureCompiler.py:388]: we have an `iterFeatureBlocks` fuction in ufo2ft.featureWriters.ast module that may help with that

File: Lib/ufo2ft/util.py [hunk 4]
```diff
@@ -612,3 +612,16 @@ def unicodeScriptExtensions(
...
+def describe_ufo(ufo: Any) -> str:
+    """Returns a description of a UFO suitable for logging."""
+    if (
+        hasattr(ufo, "reader")
+        and hasattr(ufo.reader, "fs")
+        and hasattr(ufo.reader.fs, "root_path")
+    ):
+        return ufo.reader.fs.root_path
+    elif ufo.info.familyName or ufo.info.styleName:
+        return f"{ufo.info.familyName} {ufo.info.styleName}"
```
Comment (anthrotype) [inline Lib/ufo2ft/util.py:626]:
```suggestion
        return " ".join(n for n in (ufo.info.familyName, ufo.info.styleName) if n is not None)
```

File: Lib/ufo2ft/featureCompiler.py (hunk 3 above)
Comment (madig) [reply Lib/ufo2ft/featureCompiler.py:388]: Shall I change it as well?

Comment (anthrotype) [reply Lib/ufo2ft/featureCompiler.py]: maybe not. the name `findCommentPattern` suggests it finds a comment anywhere, so let's keep it like that

File: Lib/ufo2ft/featureWriters/kernFeatureWriter.py [hunk 5]
```diff
@@ -198,6 +204,21 @@ def setContext(self, font, feaFile, compiler=None):
...
 
+        # If the font contains kerning, the feaFile contains `kern` or `dist`
+        # feature blocks, but we have no insertion markers (or they were
+        # misspelt and ignored), warn the user that the kerning blocks in the
+        # feaFile take precedence and other kerning is dropped.
+        if (
+            font.kerning
+            and ctx.existingFeatures & self.features
+            and not ctx.insertComments
+        ):
```
[the remark this answers is not in the document]
Comment (madig) [reply Lib/ufo2ft/featureWriters/kernFeatureWriter.py:215]: Hm yes. We should then actually warn if we are in append mode and have insertion markers, no? 🤔

[the remark this answers is not in the document]
Comment (anthrotype) [reply Lib/ufo2ft/featureWriters/kernFeatureWriter.py:215]: I don't think so, "append" is opt-in, so one should know what they are doing when they use it

[the remark this answers is not in the document]
Comment (madig) [reply Lib/ufo2ft/featureWriters/kernFeatureWriter.py:215]: Oh and the miscasing-search will trigger independently of writer mode 🤔 Or maybe it should only trigger if at least one writer is in skip-mode?

[the remark this answers is not in the document]
Comment (anthrotype) [reply Lib/ufo2ft/featureWriters/kernFeatureWriter.py:215]: > only trigger if at least one writer is in skip-mode?

that sounds good

**Review by anthrotype — APPROVED** [review]

**Pull request merged**